File: mobile_app/lib/licensing/offline_license.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:crypto/crypto.dart' as hashes;
import 'package:cryptography/cryptography.dart';

const _hslic1Prefix = 'HSLIC1';
const _hsreq1Prefix = 'HSREQ1';
const _hsrvl1Prefix = 'HSRVL1';
const _hslic1SignatureDomain = 'HiddenShield-Offline-License-v1';
const _hsreq1ChecksumDomain = 'HiddenShield-Offline-Activation-Request-v1';
const _hsrvl1SignatureDomain = 'HiddenShield-Offline-Revocation-List-v1';
const _installationIdDomain = 'HiddenShield-Installation-v1';
const _hslic1PayloadKeys = <String>[
  'expiresAt',
  'installationId',
  'issuedAt',
  'keyId',
  'licenseId',
  'notBefore',
  'productCode',
  'schemaVersion',
];
const _hsreq1PayloadKeys = <String>[
  'appVersion',
  'createdAt',
  'installationId',
  'nonce',
  'platform',
  'requestId',
  'requestedProductCode',
  'schemaVersion',
];
const _hsrvl1PayloadKeys = <String>[
  'generatedAt',
  'keyId',
  'listId',
  'listType',
  'revokedLicenseIds',
  'schemaVersion',
  'sequence',
];
final _timestampPattern = RegExp(r'^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z$');
final _installationIdPattern = RegExp(r'^[A-Za-z0-9_-]{43}$');
final _noncePattern = RegExp(r'^[A-Za-z0-9_-]{22}$');
final _identifierPattern = RegExp(r'^[a-z0-9][a-z0-9._-]{2,63}$');
final _appVersionPattern = RegExp(
  r'^[0-9]+\.[0-9]+\.[0-9]+(?:[-+][A-Za-z0-9.-]+)?$',
);
const _platforms = {'windows', 'macos', 'linux', 'android', 'ios'};

class OfflineLicensePayloadV1 {
  const OfflineLicensePayloadV1({
    required this.expiresAt,
    required this.installationId,
    required this.issuedAt,
    required this.keyId,
    required this.licenseId,
    required this.notBefore,
    required this.productCode,
    required this.schemaVersion,
  });

  final String expiresAt;
  final String installationId;
  final String issuedAt;
  final String keyId;
  final String licenseId;
  final String notBefore;
  final String productCode;
  final int schemaVersion;

  Map<String, Object> toJson() => {
    'expiresAt': expiresAt,
    'installationId': installationId,
    'issuedAt': issuedAt,
    'keyId': keyId,
    'licenseId': licenseId,
    'notBefore': notBefore,
    'productCode': productCode,
    'schemaVersion': schemaVersion,
  };
}

class ActivationRequestPayloadV1 {
  const ActivationRequestPayloadV1({
    required this.appVersion,
    required this.createdAt,
    required this.installationId,
    required this.nonce,
    required this.platform,
    required this.requestId,
    required this.requestedProductCode,
    required this.schemaVersion,
  });

  final String appVersion;
  final String createdAt;
  final String installationId;
  final String nonce;
  final String platform;
  final String requestId;
  final String requestedProductCode;
  final int schemaVersion;

  Map<String, Object> toJson() => {
    'appVersion': appVersion,
    'createdAt': createdAt,
    'installationId': installationId,
    'nonce': nonce,
    'platform': platform,
    'requestId': requestId,
    'requestedProductCode': requestedProductCode,
    'schemaVersion': schemaVersion,
  };
}

class RevocationListPayloadV1 {
  const RevocationListPayloadV1({
    required this.generatedAt,
    required this.keyId,
    required this.listId,
    required this.listType,
    required this.revokedLicenseIds,
    required this.schemaVersion,
    required this.sequence,
  });

  final String generatedAt;
  final String keyId;
  final String listId;
  final String listType;
  final List<String> revokedLicenseIds;
  final int schemaVersion;
  final int sequence;

  Map<String, Object> toJson() => {
    'generatedAt': generatedAt,
    'keyId': keyId,
    'listId': listId,
    'listType': listType,
    'revokedLicenseIds': revokedLicenseIds,
    'schemaVersion': schemaVersion,
    'sequence': sequence,
  };
}

class ParsedOfflineLicenseV1 {
  const ParsedOfflineLicenseV1({
    required this.payload,
    required this.payloadBytes,
    required this.signatureBytes,
    required this.signingMessage,
  });

  final OfflineLicensePayloadV1 payload;
  final Uint8List payloadBytes;
  final Uint8List signatureBytes;
  final Uint8List signingMessage;
}

class ParsedActivationRequestV1 {
  const ParsedActivationRequestV1({
    required this.payload,
    required this.payloadBytes,
    required this.checksumBytes,
  });

  final ActivationRequestPayloadV1 payload;
  final Uint8List payloadBytes;
  final Uint8List checksumBytes;
}

class ParsedRevocationListV1 {
  const ParsedRevocationListV1({
    required this.payload,
    required this.payloadBytes,
    required this.signatureBytes,
    required this.signingMessage,
  });

  final RevocationListPayloadV1 payload;
  final Uint8List payloadBytes;
  final Uint8List signatureBytes;
  final Uint8List signingMessage;
}

String encodeActivationRequestV1(ActivationRequestPayloadV1 payload) {
  final payloadText = jsonEncode(payload.toJson());
  final payloadBytes = Uint8List.fromList(utf8.encode(payloadText));
  final checksum = hashes.sha256.convert([
    ...utf8.encode('$_hsreq1ChecksumDomain\u0000'),
    ...payloadBytes,
  ]);
  return '$_hsreq1Prefix.'
      '${base64UrlEncode(payloadBytes).replaceAll('=', '')}.'
      '${base64UrlEncode(checksum.bytes.sublist(0, 12)).replaceAll('=', '')}';
}

ParsedOfflineLicenseV1 parseOfflineLicenseV1(String token) {
  final decoded = _decodeThreeSegmentToken(
    token,
    _hslic1Prefix,
    'offline_license_invalid_format',
    64,
  );
  final payloadText = _decodeUtf8(
    decoded.payloadBytes,
    'offline_license_invalid_format',
  );
  final document = _parseObject(payloadText, 'offline_license_invalid_format');
  _assertExactKeys(
    document,
    _hslic1PayloadKeys,
    'offline_license_unknown_schema',
  );
  _assertCanonical(document, payloadText);
  if (document['schemaVersion'] != 1) {
    throw const FormatException('offline_license_unknown_schema');
  }
  if (document['productCode'] != 'creator_offline') {
    throw const FormatException('offline_license_feature_profile_invalid');
  }
  final payload = OfflineLicensePayloadV1(
    expiresAt: _requiredString(document, 'expiresAt'),
    installationId: _requiredString(document, 'installationId'),
    issuedAt: _requiredString(document, 'issuedAt'),
    keyId: _requiredString(document, 'keyId'),
    licenseId: _requiredString(document, 'licenseId'),
    notBefore: _requiredString(document, 'notBefore'),
    productCode: _requiredString(document, 'productCode'),
    schemaVersion: _requiredInteger(document, 'schemaVersion'),
  );
  if (!_timestampPattern.hasMatch(payload.expiresAt) ||
      !_installationIdPattern.hasMatch(payload.installationId) ||
      !_timestampPattern.hasMatch(payload.issuedAt) ||
      !_identifierPattern.hasMatch(payload.keyId) ||
      !_identifierPattern.hasMatch(payload.licenseId) ||
      !_timestampPattern.hasMatch(payload.notBefore)) {
    throw const FormatException('offline_license_invalid_format');
  }
  return ParsedOfflineLicenseV1(
    payload: payload,
    payloadBytes: decoded.payloadBytes,
    signatureBytes: decoded.trailerBytes,
    signingMessage: _signingMessage(
      _hslic1SignatureDomain,
      decoded.payloadBytes,
    ),
  );
}

ParsedActivationRequestV1 parseActivationRequestV1(String token) {
  final decoded = _decodeThreeSegmentToken(
    token,
    _hsreq1Prefix,
    'offline_license_request_invalid_format',
    12,
  );
  final payloadText = _decodeUtf8(
    decoded.payloadBytes,
    'offline_license_request_invalid_format',
  );
  final document = _parseObject(
    payloadText,
    'offline_license_request_invalid_format',
  );
  _assertExactKeys(
    document,
    _hsreq1PayloadKeys,
    'offline_license_request_unknown_schema',
  );
  _assertCanonical(
    document,
    payloadText,
    'offline_license_request_non_canonical_payload',
  );
  if (document['schemaVersion'] != 1) {
    throw const FormatException('offline_license_request_unknown_schema');
  }
  if (document['requestedProductCode'] != 'creator_offline') {
    throw const FormatException('offline_license_request_product_invalid');
  }
  final payload = ActivationRequestPayloadV1(
    appVersion: _requiredString(document, 'appVersion'),
    createdAt: _requiredString(document, 'createdAt'),
    installationId: _requiredString(document, 'installationId'),
    nonce: _requiredString(document, 'nonce'),
    platform: _requiredString(document, 'platform'),
    requestId: _requiredString(document, 'requestId'),
    requestedProductCode: _requiredString(document, 'requestedProductCode'),
    schemaVersion: _requiredInteger(document, 'schemaVersion'),
  );
  if (!_appVersionPattern.hasMatch(payload.appVersion) ||
      !_timestampPattern.hasMatch(payload.createdAt) ||
      !_installationIdPattern.hasMatch(payload.installationId) ||
      !_noncePattern.hasMatch(payload.nonce) ||
      !_platforms.contains(payload.platform) ||
      !_identifierPattern.hasMatch(payload.requestId)) {
    throw const FormatException('offline_license_request_invalid_format');
  }
  return ParsedActivationRequestV1(
    payload: payload,
    payloadBytes: decoded.payloadBytes,
    checksumBytes: decoded.trailerBytes,
  );
}

ParsedRevocationListV1 parseRevocationListV1(String token) {
  final decoded = _decodeThreeSegmentToken(
    token,
    _hsrvl1Prefix,
    'offline_license_revocation_invalid_format',
    64,
  );
  final payloadText = _decodeUtf8(
    decoded.payloadBytes,
    'offline_license_revocation_invalid_format',
  );
  final document = _parseObject(
    payloadText,
    'offline_license_revocation_invalid_format',
  );
  _assertExactKeys(
    document,
    _hsrvl1PayloadKeys,
    'offline_license_revocation_unknown_schema',
  );
  _assertCanonical(
    document,
    payloadText,
    'offline_license_revocation_non_canonical_payload',
  );
  if (document['schemaVersion'] != 1) {
    throw const FormatException('offline_license_revocation_unknown_schema');
  }
  if (document['listType'] != 'offline_license_revocations') {
    throw const FormatException('offline_license_revocation_list_invalid');
  }
  final sequence = _requiredInteger(document, 'sequence');
  if (sequence < 1) {
    throw const FormatException('offline_license_revocation_sequence_invalid');
  }
  final revokedLicenseIds = _requiredStringList(document, 'revokedLicenseIds');
  final payload = RevocationListPayloadV1(
    generatedAt: _requiredString(document, 'generatedAt'),
    keyId: _requiredString(document, 'keyId'),
    listId: _requiredString(document, 'listId'),
    listType: _requiredString(document, 'listType'),
    revokedLicenseIds: revokedLicenseIds,
    schemaVersion: _requiredInteger(document, 'schemaVersion'),
    sequence: sequence,
  );
  if (!_timestampPattern.hasMatch(payload.generatedAt) ||
      !_identifierPattern.hasMatch(payload.keyId) ||
      !_identifierPattern.hasMatch(payload.listId) ||
      !_isSortedUniqueIdentifiers(payload.revokedLicenseIds)) {
    throw const FormatException('offline_license_revocation_list_invalid');
  }
  return ParsedRevocationListV1(
    payload: payload,
    payloadBytes: decoded.payloadBytes,
    signatureBytes: decoded.trailerBytes,
    signingMessage: _signingMessage(
      _hsrvl1SignatureDomain,
      decoded.payloadBytes,
    ),
  );
}

Future<bool> verifyOfflineLicenseV1Signature(
  ParsedOfflineLicenseV1 parsed,
  Uint8List publicKeyBytes,
) {
  return _verifyEd25519(
    parsed.signingMessage,
    parsed.signatureBytes,
    publicKeyBytes,
  );
}

bool verifyActivationRequestV1Checksum(ParsedActivationRequestV1 parsed) {
  final digest = hashes.sha256.convert(
    _signingMessage(_hsreq1ChecksumDomain, parsed.payloadBytes),
  );
  return _constantTimeEqual(
    parsed.checksumBytes,
    Uint8List.fromList(digest.bytes.take(12).toList(growable: false)),
  );
}

Future<bool> verifyRevocationListV1Signature(
  ParsedRevocationListV1 parsed,
  Uint8List publicKeyBytes,
) {
  return _verifyEd25519(
    parsed.signingMessage,
    parsed.signatureBytes,
    publicKeyBytes,
  );
}

Future<void> validateOfflineArtifactV1(
  String artifactType,
  String token, {
  Uint8List? publicKeyBytes,
}) async {
  if (artifactType == 'activation_request') {
    final parsed = parseActivationRequestV1(token);
    if (!verifyActivationRequestV1Checksum(parsed)) {
      throw const FormatException('offline_license_request_checksum_mismatch');
    }
    return;
  }
  if (publicKeyBytes == null) {
    throw const FormatException('offline_license_unknown_key');
  }
  if (artifactType == 'license') {
    final parsed = parseOfflineLicenseV1(token);
    if (!await verifyOfflineLicenseV1Signature(parsed, publicKeyBytes)) {
      throw const FormatException('offline_license_signature_invalid');
    }
    return;
  }
  if (artifactType == 'revocation_list') {
    final parsed = parseRevocationListV1(token);
    if (!await verifyRevocationListV1Signature(parsed, publicKeyBytes)) {
      throw const FormatException(
        'offline_license_revocation_signature_invalid',
      );
    }
    return;
  }
  throw const FormatException('offline_license_invalid_format');
}

String deriveInstallationIdV1(Uint8List installationSecret, Uint8List salt) {
  if (installationSecret.length != 32 || salt.length != 16) {
    throw const FormatException('offline_license_secure_storage_unavailable');
  }
  final digest = hashes.sha256.convert([
    ...utf8.encode('$_installationIdDomain\u0000'),
    ...installationSecret,
    ...salt,
  ]);
  return base64UrlEncode(digest.bytes).replaceAll('=', '');
}

Uint8List decodeOfflineLicenseBase64Url(String value) {
  return _decodeBase64UrlFor(value, 'offline_license_invalid_format');
}

Future<bool> _verifyEd25519(
  Uint8List message,
  Uint8List signatureBytes,
  Uint8List publicKeyBytes,
) async {
  if (publicKeyBytes.length != 32) {
    throw const FormatException('offline_license_unknown_key');
  }
  return Ed25519().verify(
    message,
    signature: Signature(
      signatureBytes,
      publicKey: SimplePublicKey(publicKeyBytes, type: KeyPairType.ed25519),
    ),
  );
}

_DecodedToken _decodeThreeSegmentToken(
  String token,
  String prefix,
  String errorCode,
  int trailerLength,
) {
  if (token.trim() != token || RegExp(r'\s').hasMatch(token)) {
    throw FormatException(errorCode);
  }
  final segments = token.split('.');
  if (segments.length != 3 || segments.first != prefix) {
    throw FormatException(errorCode);
  }
  final payloadBytes = _decodeBase64UrlFor(segments[1], errorCode);
  final trailerBytes = _decodeBase64UrlFor(segments[2], errorCode);
  if (trailerBytes.length != trailerLength) {
    throw FormatException(errorCode);
  }
  return _DecodedToken(payloadBytes, trailerBytes);
}

Uint8List _decodeBase64UrlFor(String value, String errorCode) {
  if (!RegExp(r'^[A-Za-z0-9_-]+$').hasMatch(value)) {
    throw FormatException(errorCode);
  }
  try {
    return Uint8List.fromList(base64Url.decode(base64Url.normalize(value)));
  } on FormatException {
    throw FormatException(errorCode);
  }
}

String _decodeUtf8(Uint8List bytes, String errorCode) {
  try {
    return utf8.decode(bytes, allowMalformed: false);
  } on FormatException {
    throw FormatException(errorCode);
  }
}

Map<String, dynamic> _parseObject(String payloadText, String errorCode) {
  try {
    final decoded = jsonDecode(payloadText);
    if (decoded is! Map<String, dynamic>) {
      throw FormatException(errorCode);
    }
    return decoded;
  } on FormatException {
    throw FormatException(errorCode);
  }
}

void _assertExactKeys(
  Map<String, dynamic> value,
  List<String> expected,
  String errorCode,
) {
  if (!_sameKeys(value.keys.toList(growable: false), expected)) {
    throw FormatException(errorCode);
  }
}

void _assertCanonical(
  Map<String, dynamic> value,
  String payloadText, [
  String errorCode = 'offline_license_non_canonical_payload',
]) {
  if (jsonEncode(value) != payloadText) {
    throw FormatException(errorCode);
  }
}

String _requiredString(Map<String, dynamic> value, String key) {
  final field = value[key];
  if (field is! String || field.isEmpty) {
    throw const FormatException('offline_license_invalid_format');
  }
  return field;
}

int _requiredInteger(Map<String, dynamic> value, String key) {
  final field = value[key];
  if (field is! int) {
    throw const FormatException('offline_license_invalid_format');
  }
  return field;
}

List<String> _requiredStringList(Map<String, dynamic> value, String key) {
  final field = value[key];
  if (field is! List || field.any((item) => item is! String)) {
    throw const FormatException('offline_license_revocation_list_invalid');
  }
  return field.cast<String>();
}

bool _isSortedUniqueIdentifiers(List<String> values) {
  String? previous;
  for (final value in values) {
    if (!_identifierPattern.hasMatch(value) ||
        (previous != null && value.compareTo(previous) <= 0)) {
      return false;
    }
    previous = value;
  }
  return true;
}

Uint8List _signingMessage(String domain, Uint8List payloadBytes) {
  return Uint8List.fromList([...utf8.encode('$domain\u0000'), ...payloadBytes]);
}

bool _constantTimeEqual(Uint8List left, Uint8List right) {
  if (left.length != right.length) return false;
  var difference = 0;
  for (var index = 0; index < left.length; index += 1) {
    difference |= left[index] ^ right[index];
  }
  return difference == 0;
}

bool _sameKeys(List<String> left, List<String> right) {
  if (left.length != right.length) return false;
  for (var index = 0; index < left.length; index += 1) {
    if (left[index] != right[index]) return false;
  }
  return true;
}

class _DecodedToken {
  const _DecodedToken(this.payloadBytes, this.trailerBytes);

  final Uint8List payloadBytes;
  final Uint8List trailerBytes;
}

File: src/lib/offline-license.ts
const HSLIC1_PREFIX = "HSLIC1";
const HSREQ1_PREFIX = "HSREQ1";
const HSRVL1_PREFIX = "HSRVL1";
const HSLIC1_SIGNATURE_DOMAIN = "HiddenShield-Offline-License-v1";
const HSREQ1_CHECKSUM_DOMAIN =
  "HiddenShield-Offline-Activation-Request-v1";
const HSRVL1_SIGNATURE_DOMAIN =
  "HiddenShield-Offline-Revocation-List-v1";
const INSTALLATION_ID_DOMAIN = "HiddenShield-Installation-v1";
const HSLIC1_PAYLOAD_KEYS = [
  "expiresAt",
  "installationId",
  "issuedAt",
  "keyId",
  "licenseId",
  "notBefore",
  "productCode",
  "schemaVersion",
] as const;
const HSREQ1_PAYLOAD_KEYS = [
  "appVersion",
  "createdAt",
  "installationId",
  "nonce",
  "platform",
  "requestId",
  "requestedProductCode",
  "schemaVersion",
] as const;
const HSRVL1_PAYLOAD_KEYS = [
  "generatedAt",
  "keyId",
  "listId",
  "listType",
  "revokedLicenseIds",
  "schemaVersion",
  "sequence",
] as const;
const TIMESTAMP_PATTERN = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z$/;
const INSTALLATION_ID_PATTERN = /^[A-Za-z0-9_-]{43}$/;
const NONCE_PATTERN = /^[A-Za-z0-9_-]{22}$/;
const IDENTIFIER_PATTERN = /^[a-z0-9][a-z0-9._-]{2,63}$/;
const APP_VERSION_PATTERN =
  /^[0-9]+\.[0-9]+\.[0-9]+(?:[-+][A-Za-z0-9.-]+)?$/;
const PLATFORMS = new Set([
  "windows",
  "macos",
  "linux",
  "android",
  "ios",
]);

export interface OfflineLicensePayloadV1 {
  expiresAt: string;
  installationId: string;
  issuedAt: string;
  keyId: string;
  licenseId: string;
  notBefore: string;
  productCode: "creator_offline";
  schemaVersion: 1;
}

export interface ActivationRequestPayloadV1 {
  appVersion: string;
  createdAt: string;
  installationId: string;
  nonce: string;
  platform: "windows" | "macos" | "linux" | "android" | "ios";
  requestId: string;
  requestedProductCode: "creator_offline";
  schemaVersion: 1;
}

export interface RevocationListPayloadV1 {
  generatedAt: string;
  keyId: string;
  listId: string;
  listType: "offline_license_revocations";
  revokedLicenseIds: string[];
  schemaVersion: 1;
  sequence: number;
}

export interface ParsedOfflineLicenseV1 {
  payload: OfflineLicensePayloadV1;
  payloadBytes: Uint8Array;
  signatureBytes: Uint8Array;
  signingMessage: Uint8Array;
}

export interface ParsedActivationRequestV1 {
  payload: ActivationRequestPayloadV1;
  payloadBytes: Uint8Array;
  checksumBytes: Uint8Array;
}

export interface ParsedRevocationListV1 {
  payload: RevocationListPayloadV1;
  payloadBytes: Uint8Array;
  signatureBytes: Uint8Array;
  signingMessage: Uint8Array;
}

export function parseOfflineLicenseV1(token: string): ParsedOfflineLicenseV1 {
  const [payloadBytes, signatureBytes] = decodeThreeSegmentToken(
    token,
    HSLIC1_PREFIX,
    "offline_license_invalid_format",
    64,
  );
  const payloadText = decodeUtf8(payloadBytes, "offline_license_invalid_format");
  const parsed = parseObject(payloadText, "offline_license_invalid_format");
  assertExactKeys(
    parsed,
    HSLIC1_PAYLOAD_KEYS,
    "offline_license_unknown_schema",
  );
  assertCanonical(parsed, payloadText);
  if (parsed.schemaVersion !== 1) {
    throw new Error("offline_license_unknown_schema");
  }
  if (parsed.productCode !== "creator_offline") {
    throw new Error("offline_license_feature_profile_invalid");
  }
  if (
    !matches(parsed.expiresAt, TIMESTAMP_PATTERN) ||
    !matches(parsed.installationId, INSTALLATION_ID_PATTERN) ||
    !matches(parsed.issuedAt, TIMESTAMP_PATTERN) ||
    !matches(parsed.keyId, IDENTIFIER_PATTERN) ||
    !matches(parsed.licenseId, IDENTIFIER_PATTERN) ||
    !matches(parsed.notBefore, TIMESTAMP_PATTERN)
  ) {
    throw new Error("offline_license_invalid_format");
  }

  return {
    payload: parsed as unknown as OfflineLicensePayloadV1,
    payloadBytes,
    signatureBytes,
    signingMessage: signingMessage(HSLIC1_SIGNATURE_DOMAIN, payloadBytes),
  };
}

export function parseActivationRequestV1(
  token: string,
): ParsedActivationRequestV1 {
  const [payloadBytes, checksumBytes] = decodeThreeSegmentToken(
    token,
    HSREQ1_PREFIX,
    "offline_license_request_invalid_format",
    12,
  );
  const payloadText = decodeUtf8(
    payloadBytes,
    "offline_license_request_invalid_format",
  );
  const parsed = parseObject(
    payloadText,
    "offline_license_request_invalid_format",
  );
  assertExactKeys(
    parsed,
    HSREQ1_PAYLOAD_KEYS,
    "offline_license_request_unknown_schema",
  );
  assertCanonical(
    parsed,
    payloadText,
    "offline_license_request_non_canonical_payload",
  );
  if (parsed.schemaVersion !== 1) {
    throw new Error("offline_license_request_unknown_schema");
  }
  if (parsed.requestedProductCode !== "creator_offline") {
    throw new Error("offline_license_request_product_invalid");
  }
  if (
    !matches(parsed.appVersion, APP_VERSION_PATTERN) ||
    !matches(parsed.createdAt, TIMESTAMP_PATTERN) ||
    !matches(parsed.installationId, INSTALLATION_ID_PATTERN) ||
    !matches(parsed.nonce, NONCE_PATTERN) ||
    typeof parsed.platform !== "string" ||
    !PLATFORMS.has(parsed.platform) ||
    !matches(parsed.requestId, IDENTIFIER_PATTERN)
  ) {
    throw new Error("offline_license_request_invalid_format");
  }

  return {
    payload: parsed as unknown as ActivationRequestPayloadV1,
    payloadBytes,
    checksumBytes,
  };
}

export function parseRevocationListV1(
  token: string,
): ParsedRevocationListV1 {
  const [payloadBytes, signatureBytes] = decodeThreeSegmentToken(
    token,
    HSRVL1_PREFIX,
    "offline_license_revocation_invalid_format",
    64,
  );
  const payloadText = decodeUtf8(
    payloadBytes,
    "offline_license_revocation_invalid_format",
  );
  const parsed = parseObject(
    payloadText,
    "offline_license_revocation_invalid_format",
  );
  assertExactKeys(
    parsed,
    HSRVL1_PAYLOAD_KEYS,
    "offline_license_revocation_unknown_schema",
  );
  assertCanonical(
    parsed,
    payloadText,
    "offline_license_revocation_non_canonical_payload",
  );
  if (parsed.schemaVersion !== 1) {
    throw new Error("offline_license_revocation_unknown_schema");
  }
  if (parsed.listType !== "offline_license_revocations") {
    throw new Error("offline_license_revocation_list_invalid");
  }
  if (!Number.isInteger(parsed.sequence) || Number(parsed.sequence) < 1) {
    throw new Error("offline_license_revocation_sequence_invalid");
  }
  if (
    !matches(parsed.generatedAt, TIMESTAMP_PATTERN) ||
    !matches(parsed.keyId, IDENTIFIER_PATTERN) ||
    !matches(parsed.listId, IDENTIFIER_PATTERN) ||
    !isSortedUniqueIdentifiers(parsed.revokedLicenseIds)
  ) {
    throw new Error("offline_license_revocation_list_invalid");
  }

  return {
    payload: parsed as unknown as RevocationListPayloadV1,
    payloadBytes,
    signatureBytes,
    signingMessage: signingMessage(HSRVL1_SIGNATURE_DOMAIN, payloadBytes),
  };
}

export async function verifyOfflineLicenseV1Signature(
  parsed: ParsedOfflineLicenseV1,
  publicKeyBytes: Uint8Array,
): Promise<boolean> {
  return verifyEd25519(
    parsed.signingMessage,
    parsed.signatureBytes,
    publicKeyBytes,
  );
}

export async function verifyActivationRequestV1Checksum(
  parsed: ParsedActivationRequestV1,
): Promise<boolean> {
  const digest = new Uint8Array(
    await crypto.subtle.digest(
      "SHA-256",
      toArrayBuffer(signingMessage(HSREQ1_CHECKSUM_DOMAIN, parsed.payloadBytes)),
    ),
  );
  return constantTimeEqual(parsed.checksumBytes, digest.slice(0, 12));
}

export async function verifyRevocationListV1Signature(
  parsed: ParsedRevocationListV1,
  publicKeyBytes: Uint8Array,
): Promise<boolean> {
  return verifyEd25519(
    parsed.signingMessage,
    parsed.signatureBytes,
    publicKeyBytes,
  );
}

export async function validateOfflineArtifactV1(
  artifactType: "license" | "activation_request" | "revocation_list",
  token: string,
  publicKeyBytes?: Uint8Array,
): Promise<void> {
  if (artifactType === "activation_request") {
    const parsed = parseActivationRequestV1(token);
    if (!(await verifyActivationRequestV1Checksum(parsed))) {
      throw new Error("offline_license_request_checksum_mismatch");
    }
    return;
  }
  if (!publicKeyBytes) {
    throw new Error("offline_license_unknown_key");
  }
  if (artifactType === "license") {
    const parsed = parseOfflineLicenseV1(token);
    if (!(await verifyOfflineLicenseV1Signature(parsed, publicKeyBytes))) {
      throw new Error("offline_license_signature_invalid");
    }
    return;
  }
  const parsed = parseRevocationListV1(token);
  if (!(await verifyRevocationListV1Signature(parsed, publicKeyBytes))) {
    throw new Error("offline_license_revocation_signature_invalid");
  }
}

export async function deriveInstallationIdV1(
  installationSecret: Uint8Array,
  salt: Uint8Array,
): Promise<string> {
  if (installationSecret.length !== 32 || salt.length !== 16) {
    throw new Error("offline_license_secure_storage_unavailable");
  }
  const digest = new Uint8Array(
    await crypto.subtle.digest(
      "SHA-256",
      toArrayBuffer(
        concatBytes(
          signingMessage(INSTALLATION_ID_DOMAIN, installationSecret),
          salt,
        ),
      ),
    ),
  );
  return encodeBase64Url(digest);
}

export function decodeBase64Url(value: string): Uint8Array {
  return decodeBase64UrlFor(value, "offline_license_invalid_format");
}

function encodeBase64Url(bytes: Uint8Array): string {
  let binary = "";
  for (const byte of bytes) binary += String.fromCharCode(byte);
  return btoa(binary)
    .replace(/\+/g, "-")
    .replace(/\//g, "_")
    .replace(/=+$/g, "");
}

function decodeThreeSegmentToken(
  token: string,
  prefix: string,
  errorCode: string,
  trailerLength: number,
): [Uint8Array, Uint8Array] {
  if (token.trim() !== token || /\s/.test(token)) {
    throw new Error(errorCode);
  }
  const segments = token.split(".");
  if (segments.length !== 3 || segments[0] !== prefix) {
    throw new Error(errorCode);
  }
  const payloadBytes = decodeBase64UrlFor(segments[1], errorCode);
  const trailerBytes = decodeBase64UrlFor(segments[2], errorCode);
  if (trailerBytes.length !== trailerLength) {
    throw new Error(errorCode);
  }
  return [payloadBytes, trailerBytes];
}

function decodeBase64UrlFor(value: string, errorCode: string): Uint8Array {
  if (!/^[A-Za-z0-9_-]+$/.test(value)) {
    throw new Error(errorCode);
  }
  try {
    const padding = "=".repeat((4 - (value.length % 4)) % 4);
    const binary = atob(
      value.replace(/-/g, "+").replace(/_/g, "/") + padding,
    );
    return Uint8Array.from(binary, (character) => character.charCodeAt(0));
  } catch {
    throw new Error(errorCode);
  }
}

function decodeUtf8(bytes: Uint8Array, errorCode: string): string {
  try {
    return new TextDecoder("utf-8", { fatal: true }).decode(bytes);
  } catch {
    throw new Error(errorCode);
  }
}

function parseObject(
  payloadText: string,
  errorCode: string,
): Record<string, unknown> {
  try {
    const parsed = JSON.parse(payloadText) as unknown;
    if (typeof parsed !== "object" || parsed === null || Array.isArray(parsed)) {
      throw new Error(errorCode);
    }
    return parsed as Record<string, unknown>;
  } catch (error) {
    if (error instanceof Error && error.message === errorCode) throw error;
    throw new Error(errorCode);
  }
}

function assertExactKeys(
  value: Record<string, unknown>,
  keys: readonly string[],
  errorCode: string,
): void {
  if (JSON.stringify(Object.keys(value)) !== JSON.stringify(keys)) {
    throw new Error(errorCode);
  }
}

function assertCanonical(
  value: Record<string, unknown>,
  payloadText: string,
  errorCode = "offline_license_non_canonical_payload",
): void {
  if (JSON.stringify(value) !== payloadText) {
    throw new Error(errorCode);
  }
}

function matches(value: unknown, pattern: RegExp): value is string {
  return typeof value === "string" && pattern.test(value);
}

function isSortedUniqueIdentifiers(value: unknown): value is string[] {
  if (!Array.isArray(value)) return false;
  let previous: string | undefined;
  for (const item of value) {
    if (!matches(item, IDENTIFIER_PATTERN)) return false;
    if (previous !== undefined && item <= previous) return false;
    previous = item;
  }
  return true;
}

function signingMessage(domain: string, payloadBytes: Uint8Array): Uint8Array {
  return concatBytes(new TextEncoder().encode(`${domain}\0`), payloadBytes);
}

async function verifyEd25519(
  message: Uint8Array,
  signatureBytes: Uint8Array,
  publicKeyBytes: Uint8Array,
): Promise<boolean> {
  if (publicKeyBytes.length !== 32) {
    throw new Error("offline_license_unknown_key");
  }
  const publicKey = await crypto.subtle.importKey(
    "raw",
    toArrayBuffer(publicKeyBytes),
    { name: "Ed25519" },
    false,
    ["verify"],
  );
  return crypto.subtle.verify(
    { name: "Ed25519" },
    publicKey,
    toArrayBuffer(signatureBytes),
    toArrayBuffer(message),
  );
}

function constantTimeEqual(left: Uint8Array, right: Uint8Array): boolean {
  if (left.length !== right.length) return false;
  let difference = 0;
  for (let index = 0; index < left.length; index += 1) {
    difference |= left[index] ^ right[index];
  }
  return difference === 0;
}

function concatBytes(left: Uint8Array, right: Uint8Array): Uint8Array {
  const output = new Uint8Array(left.length + right.length);
  output.set(left);
  output.set(right, left.length);
  return output;
}

function toArrayBuffer(bytes: Uint8Array): ArrayBuffer {
  return Uint8Array.from(bytes).buffer;
}
